dpbench/benchmarks/npbench/azimint_hist/azimint_hist_numba_dpex_p.py

```diff
@@ -0,0 +1,62 @@
+# SPDX-FileCopyrightText: 2014 Jérôme Kieffer et al.
+# SPDX-FileCopyrightText: 2021 ETH Zurich and the NPBench authors
+# SPDX-FileCopyrightText: 2022 - 2023 Intel Corporation
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""
+Jérôme Kieffer and Giannis Ashiotis. Pyfai: a python library for
+high performance azimuthal integration on gpu, 2014. In Proceedings of the
+7th European Conference on Python in Science (EuroSciPy 2014).
+"""
+
+import dpnp as np
+import numba as nb
+from numba_dpex import dpjit
+
+
+@dpjit
+def get_bin_edges_prange(a, bins):
+    bin_edges = np.zeros((bins + 1,), dtype=np.float64)
+    a_min = a.min()
+    a_max = a.max()
+    delta = (a_max - a_min) / bins
+    for i in nb.prange(bin_edges.shape[0]):
+        bin_edges[i] = a_min + i * delta
+
+    bin_edges[-1] = a_max  # Avoid roundoff error on last point
+    return bin_edges
+
+
+@dpjit
+def compute_bin(x, bin_edges):
+    # assuming uniform bins for now
+    n = bin_edges.shape[0] - 1
+    a_min = bin_edges[0]
+    a_max = bin_edges[-1]
+
+    # special case to mirror NumPy behavior for last bin
+    if x == a_max:
+        return n - 1  # a_max always in last bin
+
+    return int(n * (x - a_min) / (a_max - a_min))
+
+
+@dpjit
+def histogram_prange(a, bins, weights):
+    hist = np.zeros((bins,), dtype=a.dtype)
+    bin_edges = get_bin_edges_prange(a, bins)
+
+    for i in nb.prange(a.shape[0]):
+        bin = compute_bin(a[i], bin_edges)
+        hist[bin] += weights[i]
```

Comment:
Does this even work in numba? reduction on non-contiguous array elements.

Reply:
that's the thing, it may not be parallelizing it still runs serially.

Reply:
I think it will still run in parallel, but not as a reduction kernel. Of course, it will produce garbage results since there is a race.
I think I tried something similar for kmeans numba_npr implementation. Because of the race, resorted to using atomics operations.